backend/cn/lib/eval.ml

```diff
@@ -0,0 +1,606 @@
+module BT = BaseTypes
+module IT = IndexTerms
+
+let check_bits_bt (sgn1, sz1) (sgn2, sz2) here : (unit, string) result =
+  if Stdlib.( = ) sgn1 sgn2 then
+    Result.error ("Mismatched signs at " ^ Locations.simple_location here)
+  else if sz1 <> sz2 then
+    Result.error ("Mismatched sizes at " ^ Locations.simple_location here)
+  else
+    Ok ()
+
+
+let handle_bounds (sgn, sz) (n : Z.t) : (Z.t, string) result =
+  let min, max = BT.bits_range (sgn, sz) in
+  match sgn with
+  | Unsigned ->
+    Ok
+      (if Z.gt n max then
+         Z.rem n (Z.add max Z.one)
+       else if Z.lt n Z.zero then
+         Z.add n max
+       else
+         n)
+  | Signed ->
+    if Z.lt n min then
+      Error "signed underflow"
+    else if Z.gt n max then
+      Error "signed overflow"
+    else
+      Ok n
+
+
+let eval_num_binop
+  (eval : IT.t -> (IT.t, string) result)
+  (bt : BT.t)
+  (here : Locations.t)
+  (op : IT.binop)
+  (f : Z.t -> Z.t -> Z.t)
+  (it1 : IT.t)
+  (it2 : IT.t)
+  : (IT.t, string) result
+  =
+  let ( let@ ) = Result.bind in
+  let return = Result.ok in
+  let@ it1 = eval it1 in
+  let@ it2 = eval it2 in
+  match (it1, it2) with
+  | IT (Const (Z n1), _, _), IT (Const (Z n2), _, _) ->
+    return @@ IT.num_lit_ (f n1 n2) bt here
+  | IT (Const (Bits ((sgn1, sz1), n1)), _, _), IT (Const (Bits ((sgn2, sz2), n2)), _, _)
+    ->
+    let@ () = check_bits_bt (sgn1, sz1) (sgn2, sz2) here in
+    (match sgn1 with
+     | Unsigned ->
+       let@ res = handle_bounds (sgn1, sz1) (f n1 n2) in
+       return @@ IT.num_lit_ res bt here
+     | Signed ->
+       let@ res = handle_bounds (sgn1, sz1) (f n1 n2) in
+       return @@ IT.num_lit_ res bt here)
+  | _ -> return @@ IT.arith_binop op (it1, it2) here
```

Comment:
Integer operations in the specification language shouldn't generate over/underflow errors: all operations are total functions (as in the SMT language).